File: apps/ios/src/services/api/apiClient.ts
import { backendEnvironment } from '../../config/backendEnvironment';

export interface AuthUser {
  id: string;
  email: string;
}

export interface AuthResponse {
  user: AuthUser;
  accessToken: string;
  accessTokenExpiresAt: string;
  refreshToken: string;
}

export interface RefreshResponse {
  accessToken: string;
  accessTokenExpiresAt: string;
  refreshToken: string;
}

export type DeviceType = 'mac' | 'iphone';
export type DevicePlatform = 'ios' | 'macos';

export interface DeviceResponse {
  id: string;
  userId: string;
  deviceType: DeviceType;
  name: string;
  platform: DevicePlatform;
  publicIdentifier: string;
  lastSeenAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface RTCIceServerConfig {
  urls: string | string[];
  username?: string;
  credential?: string;
}

export interface DevicePairResponse {
  pairId: string;
  device: DeviceResponse;
}

export class APIError extends Error {
  constructor(
    message: string,
    public readonly status?: number,
  ) {
    super(message);
    this.name = 'APIError';
  }
}

/**
 * REST client for the backend's auth/devices endpoints (see
 * services/backend). Does not manage token storage or refresh scheduling —
 * callers (authStore) own that. Mirrors apps/mac/RemoteMac's APIClient.swift.
 */
class APIClient {
  constructor(private readonly baseURL: string = backendEnvironment.apiBaseURL) {}

  register(email: string, password: string): Promise<AuthResponse> {
    return this.post('/auth/register', { email, password });
  }

  login(email: string, password: string): Promise<AuthResponse> {
    return this.post('/auth/login', { email, password });
  }

  refresh(refreshToken: string): Promise<RefreshResponse> {
    return this.post('/auth/refresh', { refreshToken });
  }

  async logout(refreshToken: string): Promise<void> {
    await this.request('/auth/logout', 'POST', { refreshToken });
  }

  async registerDevice(input: {
    name: string;
    platform: DevicePlatform;
    deviceType: DeviceType;
    publicIdentifier: string;
    accessToken: string;
  }): Promise<DeviceResponse> {
    const { accessToken, ...body } = input;
    const { device } = await this.post<{ device: DeviceResponse }>('/devices', body, accessToken);
    return device;
  }

  async listDevices(accessToken: string): Promise<DeviceResponse[]> {
    const { devices } = await this.get<{ devices: DeviceResponse[] }>('/devices', accessToken);
    return devices;
  }

  async deleteDevice(deviceId: string, accessToken: string): Promise<void> {
    await this.request(`/devices/${deviceId}`, 'DELETE', undefined, accessToken);
  }

  async getIceServers(accessToken: string): Promise<RTCIceServerConfig[]> {
    const { iceServers } = await this.get<{ iceServers: RTCIceServerConfig[] }>(
      '/webrtc/ice-servers',
      accessToken,
    );
    return iceServers;
  }

  /**
   * Lets this device restore "who am I paired with" on launch, since a live
   * `pair.complete` WebSocket message only arrives while connected at the
   * moment a pairing happens.
   */
  async listDevicePairs(deviceId: string, accessToken: string): Promise<DevicePairResponse[]> {
    const { pairs } = await this.get<{ pairs: DevicePairResponse[] }>(`/devices/${deviceId}/pairs`, accessToken);
    return pairs;
  }

  private post<T>(path: string, body: Record<string, unknown>, accessToken?: string): Promise<T> {
    return this.request<T>(path, 'POST', body, accessToken);
  }

  private get<T>(path: string, accessToken?: string): Promise<T> {
    return this.request<T>(path, 'GET', undefined, accessToken);
  }

  private async request<T>(
    path: string,
    method: string,
    body?: Record<string, unknown>,
    accessToken?: string,
  ): Promise<T> {
    const headers: Record<string, string> = {};
    if (body) headers['Content-Type'] = 'application/json';
    if (accessToken) headers.Authorization = `Bearer ${accessToken}`;

    let response: Response;
    try {
      response = await fetch(`${this.baseURL}${path}`, {
        method,
        headers,
        body: body ? JSON.stringify(body) : undefined,
      });
    } catch (error) {
      throw new APIError(error instanceof Error ? error.message : 'Network request failed');
    }

    if (!response.ok) {
      const message = await response
        .json()
        .then((data: { error?: string }) => data.error)
        .catch(() => undefined);
      throw new APIError(message ?? `Request failed (${response.status})`, response.status);
    }

    if (response.status === 204) {
      return undefined as T;
    }
    return (await response.json()) as T;
  }
}

export const apiClient = new APIClient();

File: apps/mac/RemoteMac/Networking/API/APIClient.swift
import Foundation

enum APIError: Error, LocalizedError {
    case server(status: Int, message: String)
    case decoding(Error)
    case network(Error)
    case unauthorized

    var errorDescription: String? {
        switch self {
        case .server(_, let message):
            return message
        case .decoding:
            return "Received an unexpected response from the server."
        case .network(let error):
            return error.localizedDescription
        case .unauthorized:
            return "You need to sign in again."
        }
    }
}

/// REST client for the backend's auth/devices/pairing endpoints
/// (see docs: services/backend). Does not manage token storage or refresh
/// scheduling — callers (AppState) own that.
final class APIClient: @unchecked Sendable {
    private let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(baseURL: URL = BackendEnvironment.current.apiBaseURL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    func register(email: String, password: String) async throws -> AuthResponse {
        try await post("/auth/register", body: ["email": email, "password": password])
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        try await post("/auth/login", body: ["email": email, "password": password])
    }

    func refresh(refreshToken: String) async throws -> RefreshResponse {
        try await post("/auth/refresh", body: ["refreshToken": refreshToken])
    }

    func logout(refreshToken: String) async throws {
        _ = try await performRaw(makeRequest(path: "/auth/logout", method: "POST", body: ["refreshToken": refreshToken], accessToken: nil))
    }

    func registerDevice(
        name: String,
        platform: String,
        deviceType: String,
        publicIdentifier: String,
        accessToken: String
    ) async throws -> DeviceResponse {
        let wrapper: CreateDeviceResponseWrapper = try await post(
            "/devices",
            body: [
                "deviceType": deviceType,
                "name": name,
                "platform": platform,
                "publicIdentifier": publicIdentifier,
            ],
            accessToken: accessToken
        )
        return wrapper.device
    }

    func listDevices(accessToken: String) async throws -> [DeviceResponse] {
        let wrapper: ListDevicesResponseWrapper = try await get("/devices", accessToken: accessToken)
        return wrapper.devices
    }

    func requestPairingCode(macDeviceId: String, accessToken: String) async throws -> PairingCodeResponse {
        try await post("/pairing/request", body: ["macDeviceId": macDeviceId], accessToken: accessToken)
    }

    func listDevicePairs(deviceId: String, accessToken: String) async throws -> [DevicePairEntry] {
        let wrapper: ListDevicePairsResponseWrapper = try await get("/devices/\(deviceId)/pairs", accessToken: accessToken)
        return wrapper.pairs
    }

    func fetchIceServers(accessToken: String) async throws -> [IceServerConfig] {
        let wrapper: IceServersResponseWrapper = try await get("/webrtc/ice-servers", accessToken: accessToken)
        return wrapper.iceServers
    }

    // MARK: - Low-level helpers

    private func post<Response: Decodable>(
        _ path: String,
        body: [String: String],
        accessToken: String? = nil
    ) async throws -> Response {
        try await perform(makeRequest(path: path, method: "POST", body: body, accessToken: accessToken))
    }

    private func get<Response: Decodable>(_ path: String, accessToken: String? = nil) async throws -> Response {
        try await perform(makeRequest(path: path, method: "GET", body: nil, accessToken: accessToken))
    }

    private func makeRequest(path: String, method: String, body: [String: String]?, accessToken: String?) -> URLRequest {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = method
        if let body {
            request.httpBody = try? encoder.encode(body)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    private func perform<Response: Decodable>(_ request: URLRequest) async throws -> Response {
        let (data, _) = try await performRaw(request)
        do {
            return try decoder.decode(Response.self, from: data)
        } catch {
            throw APIError.decoding(error)
        }
    }

    @discardableResult
    private func performRaw(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError.network(error)
        }
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.network(URLError(.badServerResponse))
        }
        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 { throw APIError.unauthorized }
            let message = (try? decoder.decode(APIErrorBody.self, from: data))?.error
                ?? "Request failed (\(httpResponse.statusCode))"
            throw APIError.server(status: httpResponse.statusCode, message: message)
        }
        return (data, httpResponse)
    }
}
